src/main/java/core/basesyntax/service/io/impl/CsvDataReader.java

```diff
@@ -0,0 +1,39 @@
+package core.basesyntax.service.io.impl;
+
+import com.opencsv.CSVReader;
+import com.opencsv.CSVReaderBuilder;
+import core.basesyntax.model.entities.Product;
+import core.basesyntax.model.entities.exception.InvalidFileExtensionException;
+import core.basesyntax.service.io.DataReader;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.List;
+
+public class CsvDataReader<T extends Product> implements DataReader<T> {
+    private final String inputFilePath;
+
+    public CsvDataReader(String inputFilePath) {
+        this.inputFilePath = inputFilePath;
+        validateFileExtension();
+    }
+
+    @Override
+    public void validateFileExtension() {
```

Comment:
can this method be private and be called in the readData?

Reply:
No, because of the overriding. I can move it to the Validator class and add a (String path) type parameter

Reply:
can we then remove this method from the interfaces and make it private in the classes?

I think there is no need to move it to Validator - in fact you can remove the Validator class and move the checks from there to the places they belong logically